### PR TITLE
Add typing to functions related to data_vars

### PR DESCRIPTION
**Pull request by Illviljan**
A smaller PR related to trying to pass T_DuckArray's along xarray's different classes.

### CODE REVIEW

File: xarray/core/dataset.py [hunk 1]
```diff
@@ -21,7 +21,16 @@
...
-from typing import IO, TYPE_CHECKING, Any, Callable, Generic, Literal, cast, overload
+from typing import (
+    IO,
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Generic,
+    Literal,
+    cast,
+    overload,
```
Comment (headtr1ck) [inline xarray/core/dataset.py:32]:
```suggestion
    overload
```

File: xarray/core/dataset.py [hunk 2]
```diff
@@ -1324,7 +1332,7 @@ def copy(
     def _copy(
         self,
         deep: bool = False,
-        data: Mapping[Any, ArrayLike] | None = None,
```
Comment (headtr1ck) [inline xarray/core/dataset.py:1327]: I assume the ArrayLike is false?

File: xarray/core/types.py [hunk 3]
```diff
@@ -187,6 +187,9 @@ def copy(
 T_Chunks = Union[int, dict[Any, Any], Literal["auto"], None]
 T_NormalizedChunks = tuple[tuple[int, ...], ...]
 
+DataVars = Mapping[Any, Any]
```
Comment (headtr1ck) [inline xarray/core/types.py:190]: Do you think we can figure out a more precise value type?
Maybe DaCompatible or so?

**Review by headtr1ck — APPROVED** [review]
That's a good start :)

File: xarray/core/dataset.py (hunk 2 above)
Comment (Illviljan) [reply xarray/core/dataset.py:1327]: yeah, the real condition is whatever Variable._copy allows.

File: xarray/core/types.py (hunk 3 above)
Comment (Illviljan) [reply xarray/core/types.py:190]: That's the hope, but it usually opens several cans of worms when changing away from Any. 

This PR was mainly just to get _MergeResult to pass through the dataset.__init__ so I can continue in the other T_DuckArray PRs.

Comment (headtr1ck) [reply xarray/core/types.py:190]: Always the same problem....

**Pull request merged**